esp-bootloader-esp-idf/src/partitions.rs

```diff
@@ -0,0 +1,765 @@
+        u32::from_le_bytes(self.binary[8..][..4].try_into().unwrap())
+    }
+
+    /// Checks for a zero-length partition.
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    /// The label of the partition.
+    pub fn label(&self) -> &'a [u8] {
+        &self.binary[12..][..16]
+    }
+
+    /// The label of the partition as `&str`.
+    pub fn label_as_str(&self) -> &'a str {
+        let array = self.label();
+        let len = array
+            .iter()
+            .position(|b| *b == 0 || *b == 0xff)
+            .unwrap_or(array.len());
+        unsafe {
+            core::str::from_utf8_unchecked(core::slice::from_raw_parts(array.as_ptr().cast(), len))
+        }
+    }
+
+    /// Raw flags of this partition. You probably want to use
+    /// [Self::is_read_only] and [Self::is_encrypted] instead.
+    pub fn flags(&self) -> u32 {
+        u32::from_le_bytes(self.binary[28..][..4].try_into().unwrap())
+    }
+
+    /// If the partition is read only.
+    pub fn is_read_only(&self) -> bool {
+        self.flags() & 0b01 != 0
+    }
+
+    /// If the partition is encrypted.
+    pub fn is_encrypted(&self) -> bool {
+        self.flags() & 0b10 != 0
+    }
+
+    /// The partition type (type and sub-type).
+    pub fn partition_type(&self) -> PartitionType {
+        match self.raw_type() {
+            0 => PartitionType::App(self.raw_subtype().try_into().unwrap()),
+            1 => PartitionType::Data(self.raw_subtype().try_into().unwrap()),
+            2 => PartitionType::Bootloader(self.raw_subtype().try_into().unwrap()),
+            3 => PartitionType::PartitionTable(self.raw_subtype().try_into().unwrap()),
+            _ => unreachable!(),
+        }
+    }
+
+    /// Provides a "view" into the partition allowing to read/write the
+    /// partition contents by using the given [embedded_storage::Storage] and/or
+    /// [embedded_storage::ReadStorage] implementation.
+    pub fn as_embedded_storage<F>(&'a self, flash: &'a mut F) -> FlashRegion<'a, F>
+    where
+        F: embedded_storage::ReadStorage,
+    {
+        FlashRegion { raw: self, flash }
+    }
+}
+
+impl core::fmt::Debug for PartitionEntry<'_> {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        f.debug_struct("PartitionEntry")
+            .field("magic", &self.magic())
+            .field("raw_type", &self.raw_type())
+            .field("raw_subtype", &self.raw_subtype())
+            .field("offset", &self.offset())
+            .field("len", &self.len())
+            .field("label", &self.label_as_str())
+            .field("flags", &self.flags())
+            .field("is_read_only", &self.is_read_only())
+            .field("is_encrypted", &self.is_encrypted())
+            .finish()
+    }
+}
+
+#[cfg(feature = "defmt")]
+impl defmt::Format for PartitionEntry<'_> {
+    fn format(&self, fmt: defmt::Formatter) {
+        defmt::write!(
+            fmt,
+            "PartitionEntry (\
```

Comment:
Do we know how this gets printed? If the output includes a huge array of spaces maybe we should use a raw string for the format (`r#"blah"#`)

Reply:
yes it looks like this `PartitionEntry (magic = 1, raw_type = 2, raw_subtype = 3, offset = 4, len = 5, label = 6, flags = 7, is_read_only = 8, is_encrypted = 9)`

Reply:
which is what I wanted it to look like